text/0111-auto-resource-detection.md

```diff
@@ -0,0 +1,92 @@
+
+Note there are some existing implementations of this already in the SDKs (see [below](#prior-art-and-alternatives)), but nothing currently in the specification.
+
+## Explanation
+
+In order to apply auto-detected resource information to all kinds of telemetry, a user will need to configure which resource detector(s) they would like to run (e.g. AWS EC2 detector). These will be configured for each tracer and meter provider.
+
+If multiple detectors are configured, and more than one of these successfully detects a resource, the resources will be merged according to the Merge interface already defined in the specification, i.e. the earliest matched resource's attributes will take precedence. Each detector may be run in parallel, but to ensure deterministic results, the resources must be merged in the order the detectors were added. In the case of the user manually supplying resource attributes in addition to resource(s) being detected, the detected resource will be merged with the supplied resource, with the supplied resource taking precedence.
+
+A default implementation of a detector that reads resource data from the `OTEL_RESOURCE` environment variable will be included in the SDK. The environment variable will contain of a list of key value pairs, and these are expected to be represented in a format similar to the [W3C Correlation-Context](https://github.com/w3c/correlation-context/blob/master/correlation_context/HTTP_HEADER_FORMAT.md#header-value), i.e.: `key1=value1,key2=value2`. This detector must always be configured as the first detector and will always be run by default.
+
+Custom resource detectors related to specific environments (e.g. specific cloud vendors) must be implemented outside of the SDK, and users will need to import these separately.
+
+## Internal details
+
+As described above, the following will be added to the Resource SDK specification:
+
+- An interface for "detectors", to retrieve resource information, that can be supplied to tracer and meter providers
+- Specification for how to merge resources returned by the configured detectors, and with a manually supplied resource as described above
+- Details of the default "From Environment Variable" detector implementation as described above
+
+This is a relatively small proposal so is easiest to explain the details with a code example:
+
+### Usage
+
+The following example in Go creates a tracer and meter provider that uses resource information automatically detected from AWS or GCP:
+
+Assumes a dependency has been added on the `otel/api`, `otel/sdk`, `otel/awsdetector`, and `otel/gcpdetector` packages.
+
+```go
+resources := resource.Detector[]{awsdetector.EC2, gcpdetector.GCE, gcpdetector.GKE}
+tp := sdktrace.NewProvider(sdktrace.MustDetectResource(resources)) // or DetectResource (see below)
+mp := push.New(..., push.MustDetectResource(resources))
+```
```

Comment:
I think we should keep the Resource building separate from the SDKs. Instead of a `WithXXX` argument accepting detectors, why not have a resource library that simply builds a resource using the detectors you give it, and then pass the built up Resource to the SDK? IOW I'd prefer if both Tracing and Metric SDKs continue to support `WithResource`.

```
  resource := resource.Must(resource.Detect(awsdetector.EC2, gcpdetector.GCE))
  tp := sdktrace.NewProvicer(sdktrace.WithResource(resource))
```

Reply:
A related remark: caching inside the detectors sounds like more trouble than it's worth. I would advocate for users to simply evaluate the resource detectors they need and merge manually.

Reply:
I originally had something closer to your suggestion, but thought passing in detectors directly would result in a slightly simpler API for users. What's the main reason(s) for wanting resource building to be separate from the SDK? The SDKs can always support both `WithResource` & `DetectResource` (maybe having both options adds some confusion?)

I do get your point on caching though. That's definitely one issue with the current proposal.

Reply:
For reference, the removal was motivated here: https://github.com/open-telemetry/oteps/pull/111#discussion_r435617411

Reply:
OK, sounds good.

Reply:
I've actually ended up changing this to be exactly as Josh suggested. See the third bullet point in the trade off / mitigations section for reasoning